### DISCUSSION TITLE
Best way to access parent properties as child component?

### DISCUSSION THREAD

**Discussion by JaViLuMa** [solidjs/solid#1141] [Q&A] [answered]
Hello!

I am coming from React and I would like to access TITLE prop from TAB component:

```ts
// App.tsx

function App() {
  return (
    <Tabs>
      <Tab title="Lemon">Lemon is yellow</Tab>
      <Tab title="Strawberry">Strawberry is red</Tab>
      <Tab title="Pear">Pear is green</Tab>
    </Tabs>
  );
};



// Tab COMPONENT

const Tab: FC<Props> = ({ children }) => {
  return <div>{children}</div>
}
```

here, inside Tabs component (below) inside mapping function, so I can pass that title prop to TabTitle component:

```ts
// Tabs COMPONENT

type Props = {
  children: ReactElement[]
}

const Tabs: FC<Props> = ({ children }) => (
  <div>
    <ul>
      {children.map((item, index) => (
        <TabTitle key={index} title={item.props.title} />
      ))}
    </ul>
    {children}
  </div>
);



// TabTitle COMPONENT

type Props = {
  title: string
}

const TabTitle: FC<Props> = ({ title }) => (
  <li>
    <button>{title}</button>
  </li>
);
```

Is something like that possible here? If not, are there any ways I can implement this differently?

So far I have something like this:
```ts
const Tabs: Component<{ children: JSXElement[] }> = (props) => (
  <div>
    <ul>
      <For each={props.children}>
        {(tab, index) => <TabTitle title={tab.title} />}
      </For>
    </ul>
    {props.children}
  </div>
);
```

But I can't seem to figure out how to make it work.

Any help is appreciated!! :D

**Comment by ryansolid** [comment, marked as answer]
I'd probably do something like this: https://playground.solidjs.com/?hash=-1262201662&version=1.4.1. In so the parent is no responsible for rendering the children and can augment at will.

**Reply by JaViLuMa** [reply]
Thank you so much for this very helpful insight!!

I have a question though: what would type of  props in `const Tab = (props) => props;` then be? I know in this example it has TITLE and CHILDREN (title: string, children: JSXElement as part of props) but it still cannot be returned this way since it's not an valid JSX element.

**Reply by ryansolid** [reply]
Yeah you just need to force it with unknown `as unknown as JSX.Element`. I do this with Switch/Match. Better this than leaving JSX.Element as any. I've seen having to do similar things with React. Not any options here to my knowledge.

**Reply by JaViLuMa** [reply]
Thank you for this!! I thought myself about using unknown but I didn't know I can do it the way you did!!

Would the `props` as an argument of the function be of type unknow also then? It would make sense.

Also the `tab` argument in callback of `For` component is now missing a type.

This is what I would currently have: https://playground.solidjs.com/?hash=-120152212&version=1.4.1

I tried getting the `tab` to get `Tab` types by using `ComponentProps<typeof Tab>` but if I would do that then the error would appear at the `each` prop of the `For` component. Any ideas for this too??

**Reply by JaViLuMa** [reply]
Okay, quick update, I did it in the following way:

First I declared a custom type checking function:

```ts
export default (x: any): x is ITab => x.title && x.children;
```
where `ITab` is `{ title: string; children: JSXElement }`.

Then I just added that type checking where necessarry:

```ts
        <For each={toArray(c())}>
          {(tab, index) =>
            isOfTabType(tab) && (
              <TabTitle
                title={tab.title}
                index={index}
                activeTab={setSelectedTab}
              />
            )
          }
        </For>
```

Hopefully this can help someone in the future!!

**Comment by JaViLuMa** [comment]


**Reply by numfin** [reply]
nope, this solutions is bad

**Reply by JaViLuMa** [reply]
@numfin I forgot this existed!

I definitely know that now, but back then it was good enough solution for me^^

**Reply by andy-stewart-design** [reply]
For someone learning Solid, can you explain why it's a bad solution?

**Reply by numfin** [reply]
@andy-stewart-design will you pay me for knowledge?

**Reply by JaViLuMa** [reply]
@numfin he asked for an explanation, not the code, don't act like this, you are literally on a discussion page of an open source framework

**Reply by numfin** [reply]
And i said nothing about code. I asked if he willing to pay me for knowledge.

**Reply by andy-stewart-design** [reply]
I'm good, thanks for the offer, though

**Reply by JaViLuMa** [reply]
You are completely missing my point. If you are already commenting something and saying "don't do this", "don't do that" then at least explain why and not try to monetise your knowledge just because you feel like it.

This is open source, if you really want to teach something and earn something because of it, you have other platforms to do that.

**Reply by JaViLuMa** [reply]
@andy-stewart-design the code is bad because you don't really want to access the parent properties in the way it is done in my code.

I only did this because I wanted to have nice way of storing my content for specific tabs and that it is all made like 

```ts
<Tabs><Tab>...</Tab></Tabs>
```

and not
```ts
<Tabs><div>...</div><another-element-you-want-to-have-as-contnet>...</Tabs>
```

What you could do is keep a state of active index and map the children (in this case tab content) and based on the index, render that content!

**Reply by numfin** [reply]
@JaViLuMa 
I am not obliged to share my knowledge, which I have earned over a long time of work. I asked if a person who wanted to learn something was willing to share something in return for me sharing valuable knowledge with him.
He refused, which is fine. But don't blame me for anything. No one needed or asked for your opinion on how I should share my experience

**Reply by numfin** [reply]
@JaViLuMa it's also ok that  you share your bad solutions twice tho ❤️

**Reply by JaViLuMa** [reply]
> @JaViLuMa I am not obliged to share my knowledge, which I have earned over a long time of work. I asked if a person who wanted to learn something was willing to share something in return for me sharing valuable knowledge with him. He refused, which is fine. But don't blame me for anything. No one needed or asked for your opinion on how I should share my experience

Fair point, I then stand completely corrected. I just thought that since the framework itself is still fairly new and open source that we shouldn't hide knowledge behind money^^

**Reply by JaViLuMa** [reply]
> @JaViLuMa it's also ok that you share your bad solutions twice tho ❤️

Then I guess I still have more to learn :)
